### DISCUSSION TITLE
ic-certified-assets extendability experience

### DISCUSSION THREAD

**Discussion by seniorjoinu** [dfinity/sdk#2635] [Feature Requests]
Hey there.
I've just tried to use `ic-certified-assets` inside my canister, but it turns out, I can't extend it's behavior (I can't even read the state's internal data, like chunks or anything), since all the functions (including StableState methods) are private. What's the point then? 

Please reconsider this decision and make it all public or at least add some ways to extend it and observe the data.

Thanks.

**Comment by seniorjoinu** [comment]
@adamspofford-dfinity @hpeebles

**Comment by adamspofford-dfinity** [comment]
This is known and being worked on, though it's less of a priority than it could be. Marking everything `pub` wouldn't be a good solution - an API you need to be staring at the source code to understand isn't all that much more useful than copying the source code into a local module. Input on what sort of API would be the most useful would be valuable.

**Comment by seniorjoinu** [comment]
Our use-case is that we're building something like on-chain dfx. This service is able to deploy asset canisters and upload assets to it. But in order to achieve that we need to first upload these assets onto our service, and then, when the bundle is ready, re-transmit it to the target asset canister. Right now we're just using a custom solution, but it would be much better if we could use asset canister module for that, since it is a well-tested solution that comes with a lot of good things ready to use.

To achieve that we need `ic-certified-assets` to at least allow reading of uploaded assets from an outer code.
